Add showLastCharWhenObscureText

zhangwei911 opened the pull request:
*Add showLastCharWhenObscureText for TextField when obscureTextfield is set to true,it will show the last char,but in the bank app,this is not allow*

Fixes https://github.com/flutter/flutter/issues/65352

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

zhangwei911 committed:
add showLastCharWhenObscureText

zhangwei911 committed:
add showLastCharWhenObscureText

zhangwei911 committed:
add showLastCharWhenObscureText

zhangwei911 committed:
add showLastCharWhenObscureText

zhangwei911 commented:
@googlebot I signed it!

justinmc reviewed (commented):
Sorry for the slow review, this is an important bug to fix!

The main thing that this PR still needs is a test.  Could you add one to editable_text_test.dart?  There should be similar tests for obscureText that you can copy and/or modify.

Also, I want to start a discussion about the name showLastCharWhenObscureText and the use of two bools for it and obscureText.  What if instead we deprecated obscureText and created an enum with three values?  Would that make the situation any simpler?

zhangwei911 committed:
1.adjust the line breaks
2.update annotation
3.cancel indented
4.add test

zhangwei911 commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -321,7 +321,7 @@ class TextField extends StatefulWidget {
   /// must not be null.
   ///
-  /// The [textAlign], [autofocus], [obscureText], [readOnly], [autocorrect],
+  /// The [textAlign], [autofocus], [obscureText], [showLastCharWhenObscureText], [readOnly], [autocorrect],
```
i have update

zhangwei911 commented:
@justinmc like this?
```
enum Obscure {
    /// show all text
    none,
    /// obscure all text
    all,
   /// obscure all text but the last char
    default
}
```

zhangwei911 committed:
1.add obscure
2.deprecated obscureText and showLastCharWhenObscure

zhangwei911 committed:
Merge branch 'master' into master

zhangwei911 committed:
fix obscure type

zhangwei911 committed:
Update text_field.dart

justinmc commented:
@zhangwei911 Yes that's what I was talking about.  I think that approach looks good.  Can we get rid of showLastCharWhenObscureText now, or is there a reason you kept it?

Also, I think we should change the name from `obscure` to `obscureTextBehavior` and the enum would be `ObscureTextBehavior`.  I think `none` and `all` are ok, but maybe `last` should change to be `delayed`?  Since really we are obscuring everything, just at a short delay.

CC @LongCatIsLooong for a second opinion on this API.

justinmc commented:
Also FYI there is a conflict on this PR, so you'll have to merge/rebase master.

weizhang911 committed:
1.delete showLastCharWhenObscure
2.rename the enum Obscure to ObscureTextBehavior
3.rename the param obscure to obscureTextBehavior
4.rename last to delayed

zhangwei911 committed:
Merge branch 'master' into master

zhangwei911 commented:
@justinmc i have delete showLastCharWhenObscure,update enum and param name.
but flutter-build — Tree is currently broken. Please do not merge this PR unless it contains a fix for the tree.
i don`t know how to fix,its my first time to pull request

zhangwei911 committed:
Merge branch 'master' into master

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -431,6 +431,15 @@ enum TextCapitalization {
 }
 
+enum ObscureTextBehavior {
+  /// show all text
```
We need full sentences for these, so maybe change to:

Indicates all text should be shown.

Indicates all text should be obscured.

Indicates all text should be obscured, but the last character is visible for a short delay.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -431,6 +431,15 @@ enum TextCapitalization {
+  /// obscure all text
+  all,
+  /// obscure all text but the last char just at a short delay
+  delayed
```
Comma here.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
   final bool readOnly;
 
+  @Deprecated(
+    'use obscure instead.'
```
Here and below, capitalize the "u", and change "obscure" to "obscureTextBehavior".

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
+  @Deprecated(
+    'use obscure instead.'
+  )
+
```
Remove this newline.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -484,11 +495,24 @@ class TextInputConfiguration {
+    'use obscure instead.'
+  )
+
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
```
This could probably use the flutter.widgets.editableText.obscureTextBehavior template.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
+  ///
+  /// {@endtemplate}
+  final ObscureTextBehavior obscureTextBehavior;
+
```
Remove extra newline here.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
+  /// replaced by [obscuringCharacter].
+  ///
+  /// Defaults to [ObscureTextBehavior.none]. Cannot be null.
+  ///
```
Remove the newline here.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -572,6 +576,18 @@ class EditableText extends StatefulWidget {
   final bool obscureText;
+  
+  /// {@template flutter.widgets.editableText.obscureTextBehavior}
+  /// Whether or not the last character entered is shown before
+  /// being obscured.
+  /// When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the last characters in the text field is
+  /// replaced by [obscuringCharacter].
```
This is probably too specific, maybe change it to:

```
How characters in the field are obscured, if at all.

For example, a password field may want to obscure the entered text so it's not readable.

When this is set to [ObscureTextBehavior.all] or [ObscureTextBehavior.delayed], the characters in the field are replaced by [obscuringCharacter].
```

But make sure that lines don't exceed 80 characters in length.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -499,6 +502,7 @@ class EditableText extends StatefulWidget {
        assert(focusNode != null),
        assert(obscuringCharacter != null && obscuringCharacter.length == 1),
        assert(obscureText != null),
+       assert(obscureTextBehavior != null),
```
I think you can also add an assertion that both `obscureText` and `obscureTextBehavior` aren't used at the same time, to avoid confusion.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -2846,6 +2846,61 @@ void main() {
     semantics.dispose();
   });
 
+  testWidgets('password fields hide the last char when obscure is ObscureTextBehavior.all', (WidgetTester tester) async {
```
Can you also test ObscureTextBehavior.delayed?

justinmc reviewed (commented):
Some cleanup to do here from my comments, but I like this approach now.  Thank you for the PR!  Let me know if my comments don't make sense.

zhangwei911 committed:
Update editable.dart

zhangwei911 committed:
Update text_input.dart

zhangwei911 committed:
Update text_input.dart

zhangwei911 committed:
Update text_input.dart

zhangwei911 committed:
Update editable_text.dart

zhangwei911 committed:
Update editable_text_test.dart

skia-gold commented:
Gold has detected about 4 new digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/77518

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update editable.dart

skia-gold commented:
Gold has detected about 2 new digest(s) on patchset 21.
View them at https://flutter-gold.skia.org/cl/github/77518

justinmc commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -351,6 +352,7 @@ class TextField extends StatefulWidget {
     this.autofocus = false,
     this.obscuringCharacter = '•',
     this.obscureText = false,
```
I think these other instances of `obscureText` should be @Deprecated as well, right?

justinmc replied:
Also in editable_text.dart, cupertino/text_field.dart, and anywhere else.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -572,6 +575,18 @@ class EditableText extends StatefulWidget {
   final bool obscureText;
+  
+  /// {@template flutter.widgets.editableText.obscureTextBehavior}
+  /// How characters in the field are obscured, if at all.
```
There should be an empty line in between paragraphs in the docs.

```dart
/// How characters in the field are obscured, if at all.
///
```

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -2846,6 +2846,116 @@ void main() {
+
+    semantics.dispose();
+  });
+
```
Could you also test that the text in the field is immediately changed to the obscuring character when set to `all`?  Something like this but without obscuringTextCharacter: https://github.com/flutter/flutter/blob/5acf7e98f16222610a4e6bfe1e36b08a90cf53a4/packages/flutter/test/widgets/editable_text_test.dart#L2961

justinmc reviewed (commented):
Check the failing analyzer check, you've got a bunch of trailing spaces that need to be removed.

Also, the goldens are failing.  It looks like there is some layout overflow.  Maybe try running one of the failing golden tests locally and see if there is an error on the console?

weizhang911 committed:
update some files

justinmc commented:
@zhangwei911 You've got analyzer errors because of your deprecation format, could you take a look? Thanks for updating the other stuff.  https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8841034535413644464/+/u/prepare_environment/Analyze/analyze/stdout

zhangwei911 committed:
Create text_field.dart

fix deprecation notice format

zhangwei911 committed:
fix deprecation notice format

zhangwei911 committed:
fix deprecation notice format

zhangwei911 committed:
fix deprecation notice format

zhangwei911 commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -350,7 +351,12 @@ class TextField extends StatefulWidget {
     this.obscuringCharacter = '•',
+    @Deprecated(
+        'use obscureTextBehavior instead.'
+        'This feature was deprecated after v1.25.0-5.0.pre.'
```
it doesn`t work

zhangwei911 replied:
it must add 1 space in the end of the first line and use the lastest dev version

zhangwei911 committed:
fix deprecation notice format

zhangwei911 committed:
fix deprecation notice format

justinmc commented:
@zhangwei911 Can you fix the conflicts?  There are also lots of weird analyzer failures.  They'll hopefully go away when you fix the conflicts.

zhangwei911 committed:
Merge branch 'master' into master

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -142,7 +142,10 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+    @Deprecated(
+      'Use obscureTextBehavior instead. '
+      'This feature was deprecated after v2.6.0-0.0.pre.',
+    )
```
I think you meant to keep `obscureText` and add the Deprecated flag above it?

zhangwei911 replied:
yes,now it does not a break change.
i found many error files,but i does not edit it

justinmc replied:
This still looks like a problem here.  There is no obscureTextBehavior, and obscureText is removed.  obscureText should remain and obscureTextBehavior should be added, right?

justinmc reviewed (commented):
@zhangwei911 Does this successfully run for you locally?  I think there are some errors that are tripping up the analyzer.

justinmc commented:
This still seems to have some problems, does it run locally for you?

zhangwei911 committed:
Update editable_text.dart

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update text_form_field_row.dart

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update text_form_field.dart

zhangwei911 committed:
Update editable.dart

zhangwei911 committed:
Update text_input.dart

zhangwei911 committed:
Update editable_text.dart

zhangwei911 committed:
Update autofill.dart

zhangwei911 commented:
> This still seems to have some problems, does it run locally for you?

i have run locally
![ObscureTextBehavior](https://user-images.githubusercontent.com/2375753/139528001-3a7746bf-b65f-45c8-bed2-52b956a5778e.gif)
![ObscureTextBehavior1](https://user-images.githubusercontent.com/2375753/139528094-a89aa1f8-428a-4681-9994-b061cd000f7a.gif)

justinmc commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -300,10 +305,12 @@ class CupertinoTextField extends StatefulWidget {
        assert(autofocus != null),
        assert(obscuringCharacter != null && obscuringCharacter.length == 1),
-       assert(obscureText != null),
+       assert(obscureText != null || obscureTextBehavior != null),
```
We should also make sure that people don't try to use both:

```dart
assert(obscureText == null || obscureTextBehavior == null);
```

Here and everywhere else that accepts obscureTextBehavior.

justinmc commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -377,7 +393,8 @@ class CupertinoTextField extends StatefulWidget {
   /// The [autocorrect], [autofocus], [clearButtonMode], [dragStartBehavior],
-  /// [expands], [maxLengthEnforced], [obscureText], [prefixMode], [readOnly],
+  /// [expands], [maxLengthEnforced], [obscureText], [obscureTextBehavior],
+  /// [prefixMode], [readOnly],
   /// [scrollPadding], [suffixMode], [textAlign], [selectionHeightStyle],
   /// [selectionWidthStyle], and [enableSuggestions] properties must not be null.
```
Can you adjust the line length of line 397 here?

justinmc commented on packages/flutter/lib/src/material/material_state_mixin.dart:
```diff
@@ -1,167 +0,0 @@
-    super.debugFillProperties(properties);
-    properties.add(DiagnosticsProperty<Set<MaterialState>>('materialStates', materialStates, defaultValue: <MaterialState>{}));
-  }
-}
```
Why was this file removed?

justinmc commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -300,10 +305,12 @@ class CupertinoTextField extends StatefulWidget {
-       smartQuotesType = smartQuotesType ?? (obscureText ? SmartQuotesType.disabled : SmartQuotesType.enabled),
+       smartDashesType = smartDashesType ?? ((obscureText != null && obscureText) ||
+           (obscureTextBehavior != null && obscureTextBehavior != ObscureTextBehavior.none) ? SmartDashesType.disabled : SmartDashesType.enabled),
+       smartQuotesType = smartQuotesType ?? ((obscureText != null && obscureText) ||
```
Instead of checking for the existence of obscureText everywhere, what if we convert it to obscureTextBehavior in the initializers and then only use obscureTextBehavior?

```dart
obscureTextBehavior = obscureTextBehavior ?? (obscureText != null && obscureText) ? ObscureTextBehavior.delayed : ObscureTextBehavior.none,
```

Then I think you can pretty much get rid of obscureText everywhere else.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -2951,6 +2951,116 @@ void main() {
+      ),
+    ));
+
+    final String expectedValue = '•' *controller.text.length;
```
This doesn't test the delay, does it?  Maybe if you enter the text via `tester.enterText`, then check that the last character is not `"•"`, then wait a second and test that it has been obscured.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -2951,6 +2951,116 @@ void main() {
+      ),
+    ));
+
+    final String expectedValue = '•' *controller.text.length;
```
Space after the `*` here.

justinmc reviewed (commented):
Some suggestions for improving this.  There also may be errors from RenderEditable if I'm reading the code right, though I haven't tried running it again.

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update material_state_mixin.dart

zhangwei911 committed:
Update text_field.dart

zhangwei911 committed:
Update editable.dart

zhangwei911 committed:
Merge branch 'master' into master

zhangwei911 commented:
Such a simple function has changed many things for so long, but there are still various mistakes. I decided to give up. Maybe my ability is limited

justinmc commented:
@zhangwei911 Thanks for the attempt here and for getting us thinking about this.  We should have something like obscureTextBehavior in the future.  Feel free to tag me for review if you open any other PRs.
